File: examples/agents/inspect-agent-loop.py
#!/usr/bin/env python3
"""Minimal host-side agent loop for `luotsi inspect`.

The script proves the JSONL control contract without choosing an LLM framework.
It starts `luotsi inspect`, waits for text, taps it if present, captures a
screenshot, then exits. Replace `run_loop` with your own planner or model call.
"""

from __future__ import annotations

import argparse
import json
import os
import queue
import shutil
import subprocess
import sys
import threading
from dataclasses import dataclass
from pathlib import Path
from typing import Any


@dataclass
class InspectProcess:
    process: subprocess.Popen[str]
    output: "queue.Queue[str | None]"


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Drive a Luotsi inspect JSONL session.")
    parser.add_argument("--device", required=True, help="ADB serial to pass to luotsi inspect.")
    parser.add_argument("--text", default="Sign in", help="Visible text to wait for and tap.")
    parser.add_argument("--text-match", default="exact", choices=("exact", "contains"), help="Text match mode.")
    parser.add_argument("--content-description", help="Optional content description selector.")
    parser.add_argument("--resource-id", help="Optional Android resource-id selector.")
    parser.add_argument("--class-name", help="Optional Android class selector.")
    parser.add_argument("--region", type=parse_region, help="Optional selector region: left,top,right,bottom.")
    parser.add_argument("--allow-ambiguous", action="store_true", help="Let Luotsi pick the highest-ranked match if multiple elements match.")
    parser.add_argument("--artifacts", help="Artifact base directory for the inspect session.")
    parser.add_argument("--luotsi", default="luotsi", help="Luotsi executable path.")
    parser.add_argument("--timeout-sec", type=int, default=15, help="Wait timeout for visible text.")
    parser.add_argument("--screenshot-label", default="agent-loop", help="Screenshot label.")
    parser.add_argument("--no-tap", action="store_true", help="Only wait for text and capture a screenshot.")
    return parser.parse_args()


def parse_region(value: str) -> dict[str, int]:
    parts = [part.strip() for part in value.split(",")]
    if len(parts) != 4:
        raise argparse.ArgumentTypeError("region must be four integers: left,top,right,bottom")

    try:
        left, top, right, bottom = [int(part) for part in parts]
    except ValueError as exc:
        raise argparse.ArgumentTypeError("region must be four integers: left,top,right,bottom") from exc

    return {"left": left, "top": top, "right": right, "bottom": bottom}


def start_inspect(args: argparse.Namespace) -> InspectProcess:
    if args.artifacts:
        Path(args.artifacts).mkdir(parents=True, exist_ok=True)

    inspect_args = ["inspect", "--device", args.device]
    if args.artifacts:
        inspect_args.extend(["--artifacts", args.artifacts])
    command = build_spawn_command(args.luotsi, inspect_args)

    process = subprocess.Popen(
        command,
        stdin=subprocess.PIPE,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
        encoding="utf-8",
        bufsize=1,
    )
    if process.stdin is None or process.stdout is None:
        raise RuntimeError("Failed to open inspect stdin/stdout pipes.")

    output: "queue.Queue[str | None]" = queue.Queue()
    threading.Thread(target=read_lines, args=(process.stdout, output), daemon=True).start()
    if process.stderr is not None:
        threading.Thread(target=forward_stderr, args=(process.stderr,), daemon=True).start()
    return InspectProcess(process, output)


def build_spawn_command(command: str, inspect_args: list[str]) -> list[str]:
    if os.name != "nt":
        return [command, *inspect_args]

    resolved = shutil.which(command) or command
    if resolved.lower().endswith((".cmd", ".bat")):
        return [
            os.environ.get("COMSPEC", "cmd.exe"),
            "/d",
            "/v:off",
            "/c",
            quote_cmd_line([resolved, *inspect_args]),
        ]
    return [resolved, *inspect_args]


def quote_cmd_line(parts: list[str]) -> str:
    return " ".join(quote_cmd_arg(part) for part in parts)


def quote_cmd_arg(value: str) -> str:
    return '"' + value.replace("%", "%%").replace('"', '""') + '"'


def read_lines(stream: Any, output: "queue.Queue[str | None]") -> None:
    try:
        for line in stream:
            output.put(line)
    finally:
        output.put(None)


def forward_stderr(stream: Any) -> None:
    for line in stream:
        sys.stderr.write(line)


def read_event(session: InspectProcess, timeout_sec: int) -> dict[str, Any]:
    try:
        line = session.output.get(timeout=timeout_sec)
    except queue.Empty as exc:
        raise TimeoutError(f"No inspect event arrived within {timeout_sec} seconds.") from exc

    if line is None:
        raise RuntimeError("Inspect session ended before the expected event arrived.")

    event = json.loads(line)
    print(json.dumps(event, separators=(",", ":")), flush=True)
    return event


def send_command(session: InspectProcess, command_id: str, command: str, **payload: Any) -> None:
    if session.process.stdin is None:
        raise RuntimeError("Inspect stdin is closed.")

    request = {"id": command_id, "command": command, **payload}
    session.process.stdin.write(json.dumps(request, separators=(",", ":")) + "\n")
    session.process.stdin.flush()


def selector_payload(args: argparse.Namespace) -> dict[str, Any]:
    payload: dict[str, Any] = {
        "text": args.text,
        "text_match": args.text_match,
    }
    if args.content_description:
        payload["content_description"] = args.content_description
    if args.resource_id:
        payload["resource_id"] = args.resource_id
    if args.class_name:
        payload["class_name"] = args.class_name
    if args.region:
        payload["region"] = args.region
    if args.allow_ambiguous:
        payload["allow_ambiguous"] = True
    return payload


def wait_for_command_result(
    session: InspectProcess,
    command_id: str,
    timeout_sec: int,
    *,
    wait_for_settle: bool = False,
) -> dict[str, Any]:
    result: dict[str, Any] | None = None
    while True:
        event = read_event(session, timeout_sec)
        if event.get("type") in {"session_error", "protocol_error"}:
            raise RuntimeError(f"Inspect session failed: {json.dumps(event, separators=(',', ':'))}")
        if event.get("type") == "command_result" and event.get("id") == command_id:
            result = event
            if not event.get("ok") or not wait_for_settle:
                return event
        if wait_for_settle and event.get("type") == "screen_delta" and event.get("id") == command_id and result:
            return result


def wait_for_initial_state(session: InspectProcess, timeout_sec: int) -> None:
    while True:
        event = read_event(session, timeout_sec)
        if event.get("type") == "screen_snapshot":
            return
        if event.get("type") in {"session_error", "protocol_error"}:
            raise RuntimeError(f"Inspect session failed: {json.dumps(event, separators=(',', ':'))}")


def run_loop(args: argparse.Namespace) -> int:
    session = start_inspect(args)
    try:
        wait_for_initial_state(session, args.timeout_sec)

        send_command(session, "1", "wait_visible", **selector_payload(args), timeout_sec=args.timeout_sec)
        visible = wait_for_command_result(session, "1", args.timeout_sec + 5)
        if not visible.get("ok"):
            send_command(session, "2", "capture_artifacts", label="agent-loop-wait-failed")
            wait_for_command_result(session, "2", args.timeout_sec)
            return 2

        if args.no_tap:
            send_command(session, "2", "screenshot", label=args.screenshot_label)
            screenshot = wait_for_command_result(session, "2", args.timeout_sec)
            if not screenshot.get("ok"):
                return 4
            return 0

        send_command(session, "2", "tap_text", **selector_payload(args), timeout_sec=5)
        tapped = wait_for_command_result(session, "2", args.timeout_sec, wait_for_settle=True)
        if not tapped.get("ok"):
            return 3

        send_command(session, "3", "screenshot", label=args.screenshot_label)
        screenshot = wait_for_command_result(session, "3", args.timeout_sec)
        if not screenshot.get("ok"):
            return 4
        return 0
    finally:
        stop_inspect(session)


def stop_inspect(session: InspectProcess) -> None:
    if session.process.poll() is None:
        try:
            send_command(session, "exit", "exit")
        except (BrokenPipeError, RuntimeError, OSError):
            pass

    try:
        exit_code = session.process.wait(timeout=5)
    except subprocess.TimeoutExpired as exc:
        session.process.terminate()
        raise RuntimeError("Inspect process did not exit within 5 seconds.") from exc

    if exit_code != 0:
        raise RuntimeError(f"Inspect process exited with code {exit_code}.")


def write_artifact_handoff(args: argparse.Namespace) -> None:
    if not args.artifacts:
        return

    print(f"inspect-agent-loop artifacts: {args.artifacts}", file=sys.stderr)
    print(
        f"inspect-agent-loop next: luotsi replay open --last --artifacts {quote_shell_arg(args.artifacts)} --dry-run",
        file=sys.stderr,
    )


def quote_shell_arg(value: str) -> str:
    safe = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_./:=-"
    if value and all(character in safe for character in value):
        return value

    return "'" + value.replace("'", "'\"'\"'") + "'"


def main() -> int:
    args: argparse.Namespace | None = None
    try:
        args = parse_args()
        exit_code = run_loop(args)
        write_artifact_handoff(args)
        return exit_code
    except Exception as exc:  # noqa: BLE001 - example script should surface concise failures.
        print(f"inspect-agent-loop failed: {exc}", file=sys.stderr)
        if args is not None:
            write_artifact_handoff(args)
        return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/agents/inspect-agent-loop.mjs
#!/usr/bin/env node

import { spawn, spawnSync } from 'node:child_process';
import { once } from 'node:events';
import { mkdirSync } from 'node:fs';
import { extname } from 'node:path';
import { createInterface } from 'node:readline';

const usage = `Usage:
  node examples/agents/inspect-agent-loop.mjs --device <serial> [options]
  bun examples/agents/inspect-agent-loop.mjs --device <serial> [options]

Options:
  --device <serial>             ADB serial to pass to luotsi inspect.
  --text <text>                 Visible text to wait for and tap. Defaults to "Sign in".
  --text-match <mode>           Text match mode: exact or contains. Defaults to exact.
  --content-description <text>  Optional content description selector.
  --resource-id <id>            Optional Android resource-id selector.
  --class-name <name>           Optional Android class selector.
  --region <l,t,r,b>            Optional selector region for element center.
  --allow-ambiguous             Let Luotsi pick the highest-ranked match if multiple elements match.
  --artifacts <path>            Artifact base directory for the inspect session.
  --luotsi <path>               Luotsi executable path. Defaults to "luotsi".
  --timeout-sec <seconds>       Wait timeout for visible text. Defaults to 15.
  --screenshot-label <label>    Screenshot label. Defaults to "agent-loop".
  --no-tap                      Only wait for text and capture a screenshot.
`;

class JsonLineReader {
  constructor(stream) {
    this.queue = [];
    this.waiters = [];
    this.closed = false;
    this.error = null;

    const lines = createInterface({ input: stream, crlfDelay: Infinity });
    lines.on('line', (line) => this.pushLine(line));
    lines.on('close', () => this.close());
  }

  pushLine(line) {
    try {
      const event = JSON.parse(line);
      process.stdout.write(`${JSON.stringify(event)}\n`);

      const waiter = this.waiters.shift();
      if (waiter) {
        waiter.resolve(event);
      } else {
        this.queue.push(event);
      }
    } catch (error) {
      this.fail(error);
    }
  }

  next(timeoutSec) {
    if (this.queue.length > 0) {
      return Promise.resolve(this.queue.shift());
    }
    if (this.error) {
      return Promise.reject(this.error);
    }
    if (this.closed) {
      return Promise.reject(new Error('Inspect session ended before the expected event arrived.'));
    }

    return new Promise((resolve, reject) => {
      const waiter = {
        resolve: (event) => {
          clearTimeout(timer);
          resolve(event);
        },
        reject: (error) => {
          clearTimeout(timer);
          reject(error);
        },
      };
      const timer = setTimeout(() => {
        this.waiters = this.waiters.filter((candidate) => candidate !== waiter);
        reject(new Error(`No inspect event arrived within ${timeoutSec} seconds.`));
      }, timeoutSec * 1000);

      this.waiters.push(waiter);
    });
  }

  close() {
    this.closed = true;
    this.fail(this.error ?? new Error('Inspect session ended before the expected event arrived.'));
  }

  fail(error) {
    this.error = error;
    const waiters = this.waiters;
    this.waiters = [];
    for (const waiter of waiters) {
      waiter.reject(error);
    }
  }
}

function parseArgs(argv) {
  const args = {
    text: 'Sign in',
    textMatch: 'exact',
    luotsi: 'luotsi',
    timeoutSec: 15,
    screenshotLabel: 'agent-loop',
  };

  for (let index = 0; index < argv.length; index += 1) {
    const name = argv[index];
    if (name === '--help' || name === '-h') {
      process.stdout.write(usage);
      process.exit(0);
    }
    if (name === '--no-tap') {
      args.noTap = true;
      continue;
    }
    if (name === '--allow-ambiguous') {
      args.allowAmbiguous = true;
      continue;
    }

    const value = argv[index + 1];
    if (!name.startsWith('--') || value === undefined || value.startsWith('--')) {
      throw new Error(`Missing value for ${name}.\n${usage}`);
    }

    index += 1;
    switch (name) {
      case '--device':
        args.device = value;
        break;
      case '--text':
        args.text = value;
        break;
      case '--text-match':
        args.textMatch = parseMatchMode(value, '--text-match');
        break;
      case '--content-description':
        args.contentDescription = value;
        break;
      case '--resource-id':
        args.resourceId = value;
        break;
      case '--class-name':
        args.className = value;
        break;
      case '--region':
        args.region = parseRegion(value);
        break;
      case '--artifacts':
        args.artifacts = value;
        break;
      case '--luotsi':
        args.luotsi = value;
        break;
      case '--timeout-sec':
        args.timeoutSec = Number.parseInt(value, 10);
        if (!Number.isFinite(args.timeoutSec) || args.timeoutSec <= 0) {
          throw new Error('--timeout-sec must be a positive integer.');
        }
        break;
      case '--screenshot-label':
        args.screenshotLabel = value;
        break;
      default:
        throw new Error(`Unknown option: ${name}\n${usage}`);
    }
  }

  if (!args.device) {
    throw new Error(`Missing required --device.\n${usage}`);
  }

  return args;
}

function parseMatchMode(value, optionName) {
  const normalized = value.toLowerCase();
  if (normalized === 'exact' || normalized === 'contains') {
    return normalized;
  }

  throw new Error(`${optionName} must be "exact" or "contains".`);
}

function parseRegion(value) {
  const numbers = value.split(',').map((part) => Number.parseInt(part.trim(), 10));
  if (numbers.length !== 4 || numbers.some((number) => !Number.isFinite(number))) {
    throw new Error('--region must be four integers: left,top,right,bottom.');
  }

  const [left, top, right, bottom] = numbers;
  return { left, top, right, bottom };
}

function startInspect(args) {
  if (args.artifacts) {
    mkdirSync(args.artifacts, { recursive: true });
  }

  const inspectArgs = ['inspect', '--device', args.device];
  if (args.artifacts) {
    inspectArgs.push('--artifacts', args.artifacts);
  }
  const spawnTarget = buildSpawnTarget(args.luotsi, inspectArgs);

  const child = spawn(spawnTarget.command, spawnTarget.args, {
    stdio: ['pipe', 'pipe', 'pipe'],
    windowsHide: true,
  });

  const reader = new JsonLineReader(child.stdout);
  child.on('error', (error) => reader.fail(error));
  child.stdin.on('error', (error) => {
    if (error.code !== 'EPIPE') {
      reader.fail(error);
    }
  });
  child.stderr.on('data', (chunk) => process.stderr.write(chunk));

  return { child, reader };
}

function buildSpawnTarget(command, args) {
  if (process.platform !== 'win32') {
    return { command, args };
  }

  const resolvedCommand = resolveWindowsCommand(command);
  if (!isCommandShim(resolvedCommand)) {
    return { command: resolvedCommand, args };
  }

  return {
    command: process.env.ComSpec ?? 'cmd.exe',
    args: ['/d', '/v:off', '/c', quoteCmdLine([resolvedCommand, ...args])],
  };
}

function resolveWindowsCommand(command) {
  const result = spawnSync('where.exe', [command], {
    encoding: 'utf8',
    windowsHide: true,
  });
  if (result.status !== 0 || !result.stdout) {
    return command;
  }

  return result.stdout
    .split(/\r?\n/u)
    .map((line) => line.trim())
    .find(Boolean) ?? command;
}

function isCommandShim(command) {
  return ['.cmd', '.bat'].includes(extname(command).toLowerCase());
}

function quoteCmdLine(parts) {
  return parts.map(quoteCmdArg).join(' ');
}

function quoteCmdArg(value) {
  return `"${String(value).replaceAll('%', '%%').replaceAll('"', '""')}"`;
}

function sendCommand(session, id, command, payload = {}) {
  if (
    session.child.exitCode !== null ||
    session.child.signalCode !== null ||
    session.child.stdin.destroyed ||
    !session.child.stdin.writable
  ) {
    throw new Error('Inspect stdin is closed.');
  }

  session.child.stdin.write(`${JSON.stringify({ id, command, ...payload })}\n`);
}

function selectorPayload(args) {
  const payload = {
    text: args.text,
    text_match: args.textMatch,
  };
  if (args.contentDescription) {
    payload.content_description = args.contentDescription;
  }
  if (args.resourceId) {
    payload.resource_id = args.resourceId;
  }
  if (args.className) {
    payload.class_name = args.className;
  }
  if (args.region) {
    payload.region = args.region;
  }
  if (args.allowAmbiguous) {
    payload.allow_ambiguous = true;
  }

  return payload;
}

async function waitForInitialState(session, timeoutSec) {
  while (true) {
    const event = await session.reader.next(timeoutSec);
    if (event.type === 'screen_snapshot') {
      return event;
    }
    if (event.type === 'session_error' || event.type === 'protocol_error') {
      throw new Error(`Inspect session failed: ${JSON.stringify(event)}`);
    }
  }
}

async function waitForCommandResult(session, id, timeoutSec, options = {}) {
  let result = null;
  while (true) {
    const event = await session.reader.next(timeoutSec);
    if (event.type === 'session_error' || event.type === 'protocol_error') {
      throw new Error(`Inspect session failed: ${JSON.stringify(event)}`);
    }
    if (event.type === 'command_result' && event.id === id) {
      result = event;
      if (!event.ok || !options.waitForSettle) {
        return event;
      }
    }
    if (options.waitForSettle && event.type === 'screen_delta' && event.id === id && result) {
      return result;
    }
  }
}

async function stopInspect(session) {
  if (!session.child.pid) {
    return;
  }

  if (session.child.exitCode !== null || session.child.signalCode !== null) {
    validateInspectExit(session.child.exitCode, session.child.signalCode);
    return;
  }

  try {
    sendCommand(session, 'exit', 'exit');
  } catch {
    // The process may already have ended after an inspect failure.
  }

  const exit = once(session.child, 'exit');
  let timeoutId;
  const timeout = new Promise((resolve) => {
    timeoutId = setTimeout(() => {
      session.child.kill();
      resolve({ timedOut: true });
    }, 5000);
  });
  const result = await Promise.race([
    exit.then(([code, signal]) => ({ code, signal, timedOut: false })),
    timeout,
  ]);
  clearTimeout(timeoutId);

  if (result.timedOut) {
    throw new Error('Inspect process did not exit within 5 seconds.');
  }

  validateInspectExit(result.code, result.signal);
}

function validateInspectExit(code, signal) {
  if (signal) {
    throw new Error(`Inspect process exited after signal ${signal}.`);
  }
  if (code !== 0) {
    throw new Error(`Inspect process exited with code ${code}.`);
  }
}

async function runLoop(args) {
  const session = startInspect(args);

  try {
    await waitForInitialState(session, args.timeoutSec);

    sendCommand(session, '1', 'wait_visible', {
      ...selectorPayload(args),
      timeout_sec: args.timeoutSec,
    });
    const visible = await waitForCommandResult(session, '1', args.timeoutSec + 5);
    if (!visible.ok) {
      sendCommand(session, '2', 'capture_artifacts', { label: 'agent-loop-wait-failed' });
      await waitForCommandResult(session, '2', args.timeoutSec);
      return 2;
    }

    if (args.noTap) {
      sendCommand(session, '2', 'screenshot', { label: args.screenshotLabel });
      const screenshot = await waitForCommandResult(session, '2', args.timeoutSec);
      if (!screenshot.ok) {
        return 4;
      }
      return 0;
    }

    sendCommand(session, '2', 'tap_text', {
      ...selectorPayload(args),
      timeout_sec: 5,
    });
    const tapped = await waitForCommandResult(session, '2', args.timeoutSec, { waitForSettle: true });
    if (!tapped.ok) {
      return 3;
    }

    sendCommand(session, '3', 'screenshot', { label: args.screenshotLabel });
    const screenshot = await waitForCommandResult(session, '3', args.timeoutSec);
    if (!screenshot.ok) {
      return 4;
    }
    return 0;
  } finally {
    await stopInspect(session);
  }
}

function writeArtifactHandoff(args) {
  if (!args.artifacts) {
    return;
  }

  process.stderr.write(`inspect-agent-loop artifacts: ${args.artifacts}\n`);
  process.stderr.write(`inspect-agent-loop next: luotsi replay packet --last --artifacts ${quoteShellArg(args.artifacts)}\n`);
}

function quoteShellArg(value) {
  const text = String(value);
  if (/^[A-Za-z0-9_./:=-]+$/u.test(text)) {
    return text;
  }

  return `'${text.replaceAll("'", "'\"'\"'")}'`;
}

let args;
try {
  args = parseArgs(process.argv.slice(2));
  const exitCode = await runLoop(args);
  writeArtifactHandoff(args);
  process.exitCode = exitCode;
} catch (error) {
  process.stderr.write(`inspect-agent-loop failed: ${error.message}\n`);
  if (args) {
    writeArtifactHandoff(args);
  }
  process.exitCode = 1;
}
